Allow to override the Makefile's `DOCKER_MOUNT` variable

wk8 opened the pull request:
**- What I did**

This patch provides some way for non-Linux devs to work on this repo without
suffering too much from the notorious terrible I/O perf of mounted volumes (eg
on OS X).

**- How I did it**

We now allow to override the Makefile's `DOCKER_MOUNT` variable through the env
variable of the same name.

The idea here is pretty simple: I/O perf on native mounted disks
on non-Linux (notably Mac OS) is just terrible, thus making it
a real pain to develop: one has to choose between re-building
the image after every single change (eg to run a test) or just
work directly inside the same container (eg with vim, but even then
one would have to re-configure their dev container every time
it gets destroyed - containers, after all, are not supposed to
be long-lived).

Allowing to override `DOCKER_MOUNT` makes it easy for everyone
to decide what their volume/syncing strategy is; for example
one can choose to use [docker-sync](https://github.com/EugenMayer/docker-sync)

This patch won't change anything for anyone who doesn't set the `DOCKER_MOUNT`
env variable in their environment; really tried to make it as little disrputive
as possible.

**- How to verify it**

The outputs of:
```
make -n shell
BIND_DIR=. make -n shell
```
should be unchanged compared to what it was before this patch; but:
```
DOCKER_MOUNT='-v test:/test' make -n shell
```
should replace all the `-v` flags from the commands above with just
`-v test:/test`, and finally:
```
DOCKER_MOUNT=' ' make -n shell
```
should be devoid of any `-v` flags.

**- A picture of a cute animal (not mandatory but encouraged)**

![bunny](https://static.boredpanda.com/blog/wp-content/uploads/2015/09/cute-bunnies-105__605.jpg)

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/moby/moby/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "wk8/allow_arbitrary_docker_run_flags" git@github.com:wk8/moby.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

AkihiroSuda commented:
Needs rebase

wk8 committed:
Allow to override the Makefile's `DOCKER_MOUNT` variable

Through the env variable of the same name.

The idea here is pretty simple: I/O perf on native mounted disks
on non-Linux (notably Mac OS) is just terrible, thus making it
a real pain to develop: one has to choose between re-building
the image after every single change (eg to run a test) or just
work directly inside the same container (eg with vim, but even then
one would have to re-configure their dev container every time
it gets destroyed - containers, after all, are not supposed to
be long-lived).

Allowing to override `DOCKER_MOUNT` makes it easy for everyone
to decide what their volume/syncing strategy is; for example
one can choose to use [docker-sync](https://github.com/EugenMayer/docker-sync)

This patch won't change anything for anyone who doesn't
set the `DOCKER_MOUNT` env variable in their environment.

Signed-off-by: Jean Rouge <jer329@cornell.edu>

wk8 commented:
@AkihiroSuda : thanks, rebased

vdemeester reviewed (approved):
SGTM :frog:

wk8 commented:
@vdemeester : thanks for the review :) Is the next step to get another sign-off before we can merge?

yongtang reviewed (approved):
LGTM

AkihiroSuda commented:
LGTM

@thaJeztah @cpuguy83 PTAL?

wk8 commented:
@cpuguy83 honestly, I tried the caching options you mention, still got really, really poor perf.

This patch is not perfect, but at least it allows people to do their own thing if they want to/know what they're doing.

AkihiroSuda commented:
ping @cpuguy83 @tianon @thaJeztah

thaJeztah reviewed (approved):
LGTM

but perhaps would be worth discussing with @djs55 about the performance issues you're seeing
